### PR TITLE
Improve reporting for containsNoDuplicates

### LINKED ISSUE #813: improve reporting for containsDuplicates
*Platform* (all, jvm, js): all
*Extension* (none, kotlin 1.3): none

## Code related feature
```
expect(listOf(1,2,1,2,3,4,4,4)).containsNoDuplicates()
```
We would like to see the following report instead of the current one:
```
expected that subject: [1, 2, 1, 2, 3, 4, 4, 4]        (java.util.Arrays.ArrayList <114044797>)
    ◆ has not: duplicate elements
      ⚬ index 0: 1          (kotlin.Int <972231853>)
         >> duplicated by index: 2
      ⚬ index 2: 2        (kotlin.Int <525432111>)
          >> duplicated by index: 3
      ⚬ index 5: 4        (kotlin.Int <1676423953>)
          >> duplicated by index: 6, 7
```

TODO:
- adjust DefaultIterablLikeAssertions, use assertionBuilder to create the desired report
- adjust IterableExpectationsSpec accordingly

## Your first contribution?
- Write a comment `I'll work on this` if you would like to take this issue over. 
  This way we get the chance to revise the description in case things have changed in the meantime, we might give you additional hints and we can assign the task to you, so that others do not start as well.
- See [Your first code contribution](https://github.com/robstoll/atrium/blob/master/.github/CONTRIBUTING.md#your-first-code-contribution) for guidelines.  
- Do not hesitate to ask questions here or to contact us via [Atrium's slack channel](https://kotlinlang.slack.com/team/U3DE1TXKP) if you need help
  ([Invite yourself](https://slack.kotlinlang.org/) in case you do not have an account yet).

### PR DESCRIPTION
**Pull request by wordhou**
Resolves #813. I modified containsNoDuplicates from DefaultIterableLikeAssertions to describe which indices the first occurrence of an element is duplicated by.

    expect(listOf(1,2,1,2,3,4,4,4)).containsNoDuplicates()

displays the message
```
expected that subject: [1, 2, 1, 2, 3, 4, 4, 4]        (java.util.Arrays.ArrayList <1577752735>)
◆ has not: duplicate elements
  ⚬ index 0: 1        (kotlin.Int <370557051>)
      » duplicated by index: 2
  ⚬ index 1: 2        (kotlin.Int <628175509>)
      » duplicated by index: 3
  ⚬ index 5: 4        (kotlin.Int <2036589020>)
      » duplicated by index: 6, 7
```

I created a new translatable `DescriptionIterableAssertion.DUPLICATED_BY` but did not translate that into German, since I don't know any!
______________________________________
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: logic/atrium-logic-common/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultIterableLikeAssertions.kt [hunk 1]
```diff
@@ -176,10 +174,47 @@ class DefaultIterableLikeAssertions : IterableLikeAssertions {
...
-        val lookupHashSet = HashSet<E>()
-        val duplicates = createIndexAssertions(list) { (_, element) ->
-            !lookupHashSet.add(element)
-        }
+        val lookupHashMap = HashMap<E, Int>()
+        val duplicateIndices = HashMap<Int, Pair<E, MutableList<Int>>>()
+
+        list.asSequence()
+            .mapWithIndex()
+            .forEach { (index, element) ->
```
Comment (wordhou) [inline logic/atrium-logic-common/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultIterableLikeAssertions.kt:182]: Is this too obscure? Rewritten to be less Kotlin-ish:
```
                val firstOccurrence = lookupHashMap[element]
                if (firstOccurrence != null) {
                    duplicateIndices.getOrPut(firstOccurrence) {
                        Pair(element, mutableListOf<Int>())
                    }.second.add(index)
                } else {
                    lookupHashMap[element] = index
                }
```

File: translations/de_CH/atrium-translations-de_CH-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionIterableAssertion.kt [hunk 2]
```diff
@@ -45,7 +45,8 @@ enum class DescriptionIterableAssertion(override val value: String) : StringBase
     WARNING_MISMATCHES_ADDITIONAL_ENTRIES(WARNING_MISMATCHES_ADDITIONAL_ELEMENTS.getDefault()),
     NEXT_ELEMENT("ein nächstes Element"),
     NO_ELEMENTS("❗❗ kann nicht eruiert werden, leeres Iterable"),
-    DUPLICATE_ELEMENTS("doppelte Elemente")
+    DUPLICATE_ELEMENTS("doppelte Elemente"),
+    DUPLICATED_BY("(translation needed) duplicated by index: %s")
```
Comment (wordhou) [inline translations/de_CH/atrium-translations-de_CH-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionIterableAssertion.kt:49]: Translation needed :joy:

File: logic/atrium-logic-common/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultIterableLikeAssertions.kt [hunk 3]
```diff
@@ -176,10 +174,47 @@ class DefaultIterableLikeAssertions : IterableLikeAssertions {
...
+                        }
+                        .showForAnyFailure
+                        .withDescriptionAndRepresentation(
+                            TranslatableWithArgs(DescriptionIterableAssertion.INDEX, index),
+                            element
+                        )
+                        .build()
+                }
+            }
+            .toList()
```
Comment (robstoll) [inline logic/atrium-logic-common/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultIterableLikeAssertions.kt:216]: I might be wrong but here `asSequence` does not pay off as you only have one transformation, to the contrary it is an unnecessary step IMO and can be simplified to
```suggestion
        val duplicates = duplicateIndices
            .map { (index, pair) ->
                pair.let { (element, duplicateIndices) ->
                    assertionBuilder.descriptive
                        .failing
                        .withFailureHint {
                            assertionBuilder.explanatoryGroup
                                .withDefaultType
                                .withExplanatoryAssertion(
                                    TranslatableWithArgs(
                                        DescriptionIterableAssertion.DUPLICATED_BY,
                                        duplicateIndices.joinToString(", ")
                                    )
                                )
                                .build()
                        }
                        .showForAnyFailure
                        .withDescriptionAndRepresentation(
                            TranslatableWithArgs(DescriptionIterableAssertion.INDEX, index),
                            element
                        )
                        .build()
                }
            }
```

File: translations/de_CH/atrium-translations-de_CH-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionIterableAssertion.kt (hunk 2 above)
Comment (robstoll) [inline translations/de_CH/atrium-translations-de_CH-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionIterableAssertion.kt:49]:
```suggestion
    DUPLICATED_BY("(dupliziert von Index: %s")
```

**Review by robstoll — CHANGES_REQUESTED** [review]
Nice work :sunglasses: 
Just one small detail and the translation. I'll merge it myself if you have enabled the feature

**Comment by robstoll** [comment]
@wordhou thanks for your second contribution to Atrium :tada: 
How about a third one, we could use your help for other issues

And don't forget to :star: Atrium if you like it

(I have merged your PR to a separate branch as I need to forgive the backward compatibility test from failing due to a new reporting)

**Pull request merged**